src/Action.cpp

```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
     }
 }
 
-void BaseAction::complete()
-{
+void BaseAction::complete() {
     status = COMPLETED;
 }
 
-void BaseAction::error(const std::string& errorMsg)
-{
+void BaseAction::error(const std::string& errorMsg) {
     this->errorMsg = errorMsg;
     status = ERROR;
 }
 
-std::string BaseAction::getErrorMsg() const
-{
+std::string BaseAction::getErrorMsg() const {
     return errorMsg;
 }
 
-CreateUser::CreateUser(const std::string& name,
-                         const std::string& recommendationType):
-                         name(name), recommendationType(recommendationType) 
-{}
+CreateUser::CreateUser(const std::string& name, const std::string& recommendationAlg): BaseAction(),
+                                                                                        _userName(name),
+                                                                    _recommendationAlg(recommendationAlg) {}
 
-void CreateUser::act(Session& sess)
-{
-    const std::unordered_map<std::string, User*>& users = sess.getUsers();
-    if(users.find(name) != users.end())
-    {
+void CreateUser::act(Session& sess) {
+    if(sess.getUsers().count(_userName)) {
+        error(USER_EXISTS_ERR);
+        return;
+    }
+    if (_recommendationAlg == "len") {
+        sess.addToUserMap(new LengthRecommenderUser(_userName));
+    }
+    else if (_recommendationAlg == "rer") {
+        sess.addToUserMap(new RerunRecommenderUser(_userName));
+    }
+    else if (_recommendationAlg == "gen") {
+        sess.addToUserMap(new GenreRecommenderUser(_userName));
+    }
+    else {
+        error(INVALID_ALG_ERR);
+        return;
+    }
+    complete();
+}
+std::string CreateUser::toString() const {
+    return ("creating new User - " + _userName); //TODO complete due to forum answer
+}
```

Comment:
toString() isn't what the Action prints. it's the description of the Action. It's what should come up in the printActionsLog.
So here it should be `"CreateUser" + getStatusString()`

Reply:
and I was trying so hard :cry: